src/workerd/api/sockets.c++

```diff
@@ -66,15 +20,22 @@ jsg::Ref<Socket> connectImplNoOutputLock(
   // TODO: Validate `address` is well formed. Do I need to use `parseAddress` here or is there
   // a better way?
 
-  // Set up the connection. This is similar to HttpClientAdapter::connect.
+  // Set up the connection.
   auto headers = kj::heap<kj::HttpHeaders>(ioContext.getHeaderTable());
-  auto paf = kj::newPromiseAndFulfiller<kj::HttpClient::ConnectResponse>();
-  auto tunnel = kj::refcounted<TCPConnectResponseImpl>(kj::mv(paf.fulfiller));
-  auto promise = client->connect(address, *headers, *tunnel)
-      .attach(kj::str(address), kj::mv(headers));
-  tunnel->setPromise(kj::mv(promise));
-  kj::Promise<kj::HttpClient::ConnectResponse> connResponse = paf.promise.attach(kj::mv(tunnel));
-  return jsg::alloc<Socket>(kj::mv(connResponse));
+  auto httpClient = kj::newHttpClient(*client);
+  auto request = httpClient->connect(address, *headers);
+  auto revocable = kj::heap<RevocableIoStream>(*request.connection);
+
+  auto promise = request.status.then([&revocable](auto status) {
+    if (status.statusCode < 200 || status.statusCode >= 300) {
+      // The connection request failed!
+      // TODO(later): The status may include an errorBody that we currently do nothing with.
+      revocable->revoke(KJ_EXCEPTION(FAILED,
```

Comment:
Doesn't the underlying stream already throw an exception in this scenario? Does your exception here end up racing with the one thrown by the underlying stream? If the underlying stream wins, will the application get an "internal error" instead of the transparent error specified here?

Reply:
It seems like what you really need here is a way to customize the exception thrown by the inner stream in this scenario. Maybe a new method added to `kj::HttpClientErrorHandler`?

Reply:
As a short-term solution since I know we're eager to get this landed, maybe it makes sense to just let it fail with "internal error" for now. We can go back and improve the error handling in a later PR. But I think the use of RevocableIoStream can be removed from this PR.

Reply:
Given that sockets is still a WIP, I'd really prefer if we could just leave a TODO here to improve this so that we don't block landing the pipelining connect any further.